sandbox/plc4c/spi/src/system.c

```diff
@@ -204,146 +204,71 @@ void plc4c_system_shutdown(plc4c_system *system) {}
-        // doesn't have one and the remaining part is simply the transport
-        // connect information.
-        if (num_question_marks == 0) {
-          transport_connect_information =
-              malloc(sizeof(char) * ((i - start_segment_index)) + 2);
-          *(transport_connect_information + (i - start_segment_index) + 1) = '\0';
-          strncpy(transport_connect_information, start_segment,
-                  (i - start_segment_index) + 1);
-        }
-        // I a question-mark was found, this is the parameters section.
-        else {
-          parameters = malloc(sizeof(char) * (i - start_segment_index) + 2);
-          *(parameters + (i - start_segment_index) + 1) = '\0';
-          strncpy(parameters, start_segment, (i - start_segment_index) + 1);
-        }
-      }
-    }
-  }
-  if (num_colons == 0) {
-    return INVALID_CONNECTION_STRING;
-  }
+  // Make a local copy so not to mess up oringal argument, needed later?
+  strcpy(connection_to_tokenize, connection_string);
 
-  // Initialize a new connection data-structure with the parsed information.
-  plc4c_connection *new_connection = malloc(sizeof(plc4c_connection));
+  // Inisalise the connection and set full connection string argument
+  new_connection = malloc(sizeof(plc4c_connection));
   plc4c_connection_initialize(new_connection);
   plc4c_connection_set_connection_string(new_connection, connection_string);
-  plc4c_connection_set_protocol_code(new_connection, protocol_code);
-  plc4c_connection_set_transport_code(new_connection, transport_code);
-  plc4c_connection_set_transport_connect_information(
-      new_connection, transport_connect_information);
-  plc4c_connection_set_parameters(new_connection, parameters);
 
+  // Get the protocol code (1st item, ':' delimited, required)
+  connection_token = strtok_r(connection_to_tokenize, ":", &connection_string_pos);
+  plc4c_connection_set_protocol_code(new_connection, connection_token);
+
+  // As protocol code required, check we found a ':'
+  if (strlen(connection_string_pos) == 0) 
+    return INVALID_CONNECTION_STRING;
+
+  // Check if protocol is followed by '//', if by just one '/' return error
+  if (strncmp(connection_string_pos, "//", 2) == 0) {
+    plc4c_connection_set_transport_code(new_connection, NULL);
+    mode = SEARCHING_FOR_QUESTION_MARKS;
+  } else if (strncmp(connection_string_pos, "/", 1) == 0) {
+    return INVALID_CONNECTION_STRING;
+  }
+
+  // Get the transport code (2nd item, ':' delimited, optional)
+  if (mode == SEARCHING_FOR_COLONS) {
+    connection_token = strtok_r(connection_string_pos, ":", &connection_string_pos);
+    plc4c_connection_set_transport_code(new_connection, connection_token);
+    if (strncmp(connection_string_pos, "//", 2)) 
+      return INVALID_CONNECTION_STRING;
+  }
+
+  // Skip over the '//' we have now asserted MUST exist, and search for
+  // connection parameters (last item, '?' delimited, optional).
+  connection_string_pos += 2;
```

Comment:
again NULL checks

Reply:
plc4c_connection_set_transport_code and similar "set" functions check for null hence I didn't, but actually yes if it reutns null subsquency opeations should not be performed as the last example shows